File: scrape.py
import requests
import re
from bs4 import BeautifulSoup

URL = "https://web6.karlsruhe.de/service/abfall/akal/akal.php"
DATE = re.compile(r"\d\d\.\d\d\.2021")


class StreetQuery:
    def __init__(self, tag):
        self.street = tag.string.strip()
        self.from_letter = self.street[0]
        self.value = int(tag["value"])

        if self.from_letter == "Z":
            self.to_letter = "["
        else:
            self.to_letter = chr(ord(self.from_letter) + 1)

    @property
    def params(self):
        return dict(von=self.from_letter, bis=self.to_letter)

    @property
    def data(self):
        return dict(anzeigen="anzeigen", strasse=self.value, hausnr="")


def get_street_queries():
    page = requests.post(URL, params=dict(von="A", bis="[")).text
    soup = BeautifulSoup(page, features='lxml')
    return [StreetQuery(tag) for tag in soup.body.find("select")]


def get_date(query):
    page = requests.post(URL, params=query.params, data=query.data).text
    soup = BeautifulSoup(page, features="lxml")
    rows = soup.body.find("div", attrs={"id": "foo"}).table.find_all("tr")
    td = rows[-1].find_all("td")[-2]
    match = DATE.search(td.text)

    if match:
        return match.group(0)

    return None


for query in get_street_queries():
    date = get_date(query)

    if date:
        print(f"{query.street};{date}")

File: src/scrape.rs
use anyhow::{anyhow, Result};
use regex::Regex;
use select::document::Document;
use select::node::Node;
use select::predicate::{Attr, Class, Name, Predicate};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct Pickup {
    pub street: String,
    pub date: String,
}

#[derive(Debug)]
pub struct Query {
    pub street: String,
    pub value: String,
}

impl Query {
    fn from(node: Node) -> Option<Query> {
        if let Some(value) = node.attr("value") {
            return Some(Query {
                street: node.text().trim().to_string(),
                value: value.to_string(),
            });
        }
        None
    }
}

pub struct Client {
    client: reqwest::Client,
    url: &'static str,
    date_expr: Regex,
}

impl Client {
    pub fn new() -> Result<Self> {
        Ok(Self {
            client: reqwest::Client::new(),
            url: "https://web6.karlsruhe.de/service/abfall/akal/akal.php",
            date_expr: Regex::new(r"(\d\d\.\d\d\.2022)")?,
        })
    }

    pub async fn queries(&self) -> Result<Vec<Query>> {
        let text = self
            .client
            .post(self.url)
            .query(&[("von", "A"), ("bis", "[")])
            .send()
            .await?
            .text()
            .await?;

        let document = Document::from(text.as_str());

        Ok(document
            .find(Class("tab_body").descendant(Name("option")))
            .into_iter()
            .filter_map(Query::from)
            .collect::<Vec<_>>())
    }

    pub async fn get_pickup(&self, query: Query) -> Result<Pickup> {
        let data = [
            ("anzeigen", "anzeigen"),
            ("strasse", &query.value),
            ("hausnr", "1"),
        ];

        let text = self
            .client
            .post(self.url)
            .query(&[("von", "A"), ("bis", "[")])
            .form(&data)
            .send()
            .await?
            .text()
            .await?;

        let node = Document::from(text.as_str())
            .find(Attr("id", "foo").descendant(Name("tbody")))
            .next()
            .ok_or_else(|| anyhow!("<div id='foo'><tbody> not found"))?
            .last_child() // last <tr>
            .ok_or_else(|| anyhow!("<tr> not found"))?
            .find(Name("td"))
            .nth(2)
            .ok_or_else(|| anyhow!("third <td> not found"))?
            .text();

        Ok(Pickup {
            date: self
                .date_expr
                .captures(&node)
                .ok_or_else(|| anyhow!("No date found for {}", query.street))?
                .get(0)
                .ok_or_else(|| anyhow!("foo"))?
                .as_str()
                .to_owned(),
            street: query.street,
        })
    }
}
